### ISSUE TITLE
Unable to read Kotlin metadata due to unsupported metadata kind: SyntheticClass on DefaultImpls

### ISSUE THREAD

**Issue by TWiStErRob** [google/dagger#3884]
https://github.com/google/dagger/issues/2379 regressed as a warning, e.g. in 2.46.1.

@ZacSweers originally reported at https://github.com/google/dagger/issues/3701#issuecomment-1530193557
@danysantiago explained it here https://github.com/google/dagger/issues/3701#issuecomment-1531702574

Opening to track this, as there should be no warnings when there are no problems.
Especially from a seemingly Java file in a Kotlin-only project.

## Mentioned Repro

By @danysantiago https://github.com/google/dagger/issues/2379#issuecomment-778263561 produces a failure along with the warning:
```
> Task :kaptKotlin FAILED
build\tmp\kapt3\stubs\main\com\example\MyModule.java:12: warning: Unable to read Kotlin metadata due to unsupported metadata kind: dagger.spi.shaded.kotlinx.metadata.jvm.KotlinClassMetadata$SyntheticClass@126ad42.
    public static final class DefaultImpls {
                        ^
build\tmp\kapt3\stubs\main\com\example\MyModule.java:9: error: @Provides methods cannot be abstract
    public abstract java.lang.String provide();
                                     ^
build\tmp\kapt3\stubs\main\com\example\MyModule.java:16: error: @Provides methods can only be present within a @Module or @ProducerModule
        public static java.lang.String provide(@org.jetbrains.annotations.NotNull
                                       ^
```

## Clean Repro

* Check out https://github.com/TWiStErRob/repros/tree/main/dagger/metadata-defaultimpls
* `gradlew build --rerun-tasks`

The one I found doesn't have default implementations (the DefaultImpls class is actually empty), and there's no other warning or error:
```
> Task :kaptKotlin
build\tmp\kapt3\stubs\main\com\example\QuickbookModule.java:21: warning: Unable to read Kotlin metadata due to unsupported metadata kind: dagger.spi.shaded.kotlinx.metadata.jvm.KotlinClassMetadata$SyntheticClass@464e91b1.
        public static final class DefaultImpls {
                            ^
```
The cause is the default argument in `@BindsInstance`:
```kotlin
@Component.Builder
interface Builder {
	@BindsInstance
	fun instance(param: String = "default"): Builder
```

**Comment by danysantiago** [comment]
Fixed in https://github.com/androidx/androidx/commit/8afd26955f9c0e9a0229c09a259999167cb2e4e3

**Comment by TWiStErRob** [comment]
Nice fix, I'm a bit confused though, how come a pure dagger issue (see [repro](https://github.com/TWiStErRob/repros/blob/main/dagger/metadata-defaultimpls/build.gradle.kts#L13-L14)) being fixed by a commit in room processor?

**Comment by bcorso** [comment]
Dagger's annotation processor uses Room's [XProcessing library](https://github.com/androidx/androidx/tree/androidx-main/room/room-compiler-processing) which is an abstraction layer used to support both Javac/KAPT and KSP. It lives in Room at the moment only because that's where the project started, but it's a general library that's not really specific to Room. At some point, it will likely be moved out of Room and located somewhere more general.

**Comment by TWiStErRob** [comment]
Aah, 🙏 thanks for the history!

**Comment by jarroyoesp** [comment]
Hi guys, in our project we are facing a similar issue, it always happens in `[PagingSources](https://developer.android.com/jetpack/androidx/releases/paging)`:
```
> Task :modules:feature-people:kaptReleaseKotlin
/.../SamplePagingSource.java:88: warning: Unable to read Kotlin metadata due to unsupported metadata kind: dagger.spi.shaded.kotlinx.metadata.jvm.KotlinClassMetadata$SyntheticClass@6adfaf9.
        public static final class DefaultImpls {
```

Could you tell us in which `Hilt` version is fixed or will be fixed? Currently we are using `2.46.1`, and just in case it helps, downgrading to `2.46`the issue dissapears
